modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C

```diff
@@ -205,15 +218,70 @@ ComputeMultiPlasticityStress::postReturnMap()
+ComputeMultiPlasticityStress::quickStep(const RankTwoTensor & stress_old, RankTwoTensor & stress, const std::vector<Real> & intnl_old,
+                                        std::vector<Real> & intnl, std::vector<Real> & pm, std::vector<Real> & cumulative_pm,
+                                        const RankTwoTensor & plastic_strain_old, RankTwoTensor & plastic_strain, const RankFourTensor & E_ijkl,
+                                        const RankTwoTensor & strain_increment, std::vector<Real> & yf, unsigned int & iterations,
+                                        RankFourTensor & consistent_tangent_operator, const std::string & called_from, const bool & final_step)
 {
-  stress = stress_old + E_ijkl*strain_increment;
-  plastic_strain = plastic_strain_old;
-  for (unsigned model = 0 ; model < _num_models ; ++model)
-    intnl[model] = intnl_old[model];
   iterations = 0;
-  consistent_tangent_operator = E_ijkl;
-  return checkAdmissible(stress, intnl, yf);
+
+  unsigned num_plastic_returns;
+  RankTwoTensor delta_dp;
+
+
+  // the following does the customized returnMap algorithm
+  // for all the plastic models.
+  unsigned custom_model = 0;
+  bool successful_return = returnMapAll(stress_old + E_ijkl*strain_increment, intnl_old, E_ijkl, _epp_tol, stress, intnl, pm, cumulative_pm, delta_dp, yf, num_plastic_returns, custom_model);
+
+
+  // the following updates the plastic_strain, when necessary
+  // and calculates the consistent_tangent_operator, when necessary
+  if (num_plastic_returns == 0)
+  {
+    // if successful_return = true, then a purely elastic step
+    // if successful_return = false, then >=1 plastic model is in
+    //    inadmissible zone and failed to return using its customized
+    //    returnMap function.
+    // In either case:
+    plastic_strain = plastic_strain_old;
+    if (successful_return && final_step)
+    {
+      if (called_from.compare("computeQpStress") == 0)
+        consistent_tangent_operator = E_ijkl;
+      else // cannot necessarily use E_ijkl since different plastic models may have been active during other substeps
+        consistent_tangent_operator = consistentTangentOperator(stress, intnl, E_ijkl, pm, cumulative_pm);
+    }
+    return successful_return;
+  }
+  else if (num_plastic_returns == 1 && successful_return)
+  {
+    // one model has successfully completed its custom returnMap algorithm
+    // and the other models have signalled they are elastic at
+    // the trial stress
+    plastic_strain = plastic_strain_old + delta_dp;
+    if (final_step)
+    {
+      if (called_from.compare("computeQpStress") == 0)
```

Comment:
A string comparison is likely to be a significant slow down here. I suggest using an enum if you must have a construct like this.